### PR TITLE
intake regionmask example

### LINKED ISSUE #93: Add new regions from remote url locations via yaml
https://www.marineregions.org/downloads.php

Implement with caching of download. See other issue.

### PR DESCRIPTION
**Pull request by aaronspring**
- [x] Closes #93 
 - [x] Tests added
 - [x] Passes `isort . && black . && flake8`
 - [x] Fully documented, including `whats-new.rst`

~based on `.rst` or should I rather use notebook?~

### CODE REVIEW

**Comment by aaronspring** [comment]
@mathause do you also have docs CI? and contributing page with your workflow would help.

I get errors in regionmask-docs after conda env update -f environment_rtd.yml --prune.

```
(regionmask-docs) aaron.spring@Aarons-MBP:~/Coding/regionmask$ pip install -e ../.
(regionmask-docs) aaron.spring@Aarons-MBP:~/Coding/regionmask$ cd docs/
(regionmask-docs) aaron.spring@Aarons-MBP:~/Coding/regionmask/docs$ make html
sphinx-build -b html -d _build/doctrees   . _build/html
Running Sphinx v2.1.2

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/Users/aaron.spring/anaconda3/envs/regionmask-docs/lib/python3.7/site-packages/sphinx/config.py", line 361, in eval_config_file
    execfile_(filename, namespace)
  File "/Users/aaron.spring/anaconda3/envs/regionmask-docs/lib/python3.7/site-packages/sphinx/util/pycompat.py", line 86, in execfile_
    exec(code, _globals)
  File "/Users/aaron.spring/Coding/regionmask/docs/conf.py", line 20, in <module>
    import regionmask
  File "/Users/aaron.spring/Coding/regionmask/regionmask/__init__.py", line 5, in <module>
    from . import core, defined_regions
  File "/Users/aaron.spring/Coding/regionmask/regionmask/defined_regions/__init__.py", line 3, in <module>
    from . import _ressources
  File "/Users/aaron.spring/Coding/regionmask/regionmask/defined_regions/_ressources.py", line 1, in <module>
    import geopandas as gp
  File "/Users/aaron.spring/anaconda3/envs/regionmask-docs/lib/python3.7/site-packages/geopandas/__init__.py", line 1, in <module>
    from geopandas.geoseries import GeoSeries
  File "/Users/aaron.spring/anaconda3/envs/regionmask-docs/lib/python3.7/site-packages/geopandas/geoseries.py", line 13, in <module>
    from geopandas.base import GeoPandasBase, _unary_op, _CoordinateIndexer
  File "/Users/aaron.spring/anaconda3/envs/regionmask-docs/lib/python3.7/site-packages/geopandas/base.py", line 6, in <module>
    from pandas.core.indexing import _NDFrameIndexer
ImportError: cannot import name '_NDFrameIndexer' from 'pandas.core.indexing' (/Users/aaron.spring/anaconda3/envs/regionmask-docs/lib/python3.7/site-packages/pandas/core/indexing.py)

make: *** [html] Error 2
```

**Comment by mathause** [comment]
Thanks! Looks good at first glance. The error does not look familiar - could be an incompatibility between geopandas and pandas versions. I need to look into this. The IPCC deadline is approaching so you may need some patience.

File: docs/intake_regionmask.rst [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+    # Use defaults so we don't get gridlines in generated docs
+    import matplotlib as mpl
+    mpl.rcdefaults()
+    mpl.use('Agg')
+
+    # cut border when saving (for maps)
+    mpl.rcParams["savefig.bbox"] = "tight"
+
+You may need to install intake_geopandas, which combines geopandas and intake, see
```
Comment (mathause) [inline docs/intake_regionmask.rst:21]:
```suggestion
You need to install intake_geopandas, which combines geopandas and intake, see
```

**Review by mathause — COMMENTED** [review]
Yes a contributors guide would be nice but quite some work. What I do is

```bash
cd docs
mamba env create -f environment_rtd.yml
make html
```
I mainly use pip in the yml file because rtd was very restrictive concerning the available ressources. I should try conda again some time...

---

Looks pretty good. I left some comments below.

Edit: and thanks for you patience!

**Comment by mathause** [comment]
Could you also add cache/ & test_cat.yml to .gitignore

**Comment by mathause** [comment]
Oh, the error you get does tell me something. Most likely that is a newer pandas version with an older geopandas version. I think `NDFrame` is a n-D pandas structure that was deprecated in pandas favour of xarray.

File: docs/index.rst [hunk 2]
```diff
@@ -70,6 +70,7 @@ Documentation
    defined_landmask
    defined_ocean_basins
    defined_scientific
+   intake_regionmask
```
[the remark this answers is not in the document]
Comment (aaronspring) [reply docs/index.rst:73]: probably. its still a way to access regions from a shp file, but its clearly not a regionmask region defined inside regionmask. moved to tutorials

File: docs/intake_regionmask.rst [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+======================
+
+Let's explore the Marine Ecoregions Of the World (MEOW) data set, which is a
+biogeographic classification of the world's coasts and shelves.
+To create a catalog we use the syntax described in
+`intake <https://intake.readthedocs.io/en/latest/catalog.html#yaml-format>`_.
+
+.. ipython:: python
+
+    %%writefile test_cat.yml
```
Comment (aaronspring) [inline docs/intake_regionmask.rst:52]: this is another problem. these magic commands work in notebooks. But I couldnt find a way how to write a new file here in the rst file. is bash somehow possible?

```
#!/bin/bash

cat > FILE.txt <<EOF

info code info 
info code info
info code info

EOF 
```

Comment (aaronspring) [reply docs/intake_regionmask.rst]: should I rather provide a notebook? new nice for git but at least this should work...

Comment (aaronspring) [reply docs/intake_regionmask.rst]: not the problem. problem caused here from remote: 
```python
>>> import intake
>>> cat = intake.open_catalog('https://raw.githubusercontent.com/aaronspring/remote_climate_data/master/master.yaml')
Traceback (most recent call last):
```
I will have to fix that. thought this was very stable but isnt.

Comment (aaronspring) [reply docs/intake_regionmask.rst]: works now.

File: docs/intake_regionmask.rst [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    url = 'https://raw.githubusercontent.com/aaronspring/remote_climate_data/master/master.yaml'
+    cat = intake.open_catalog(url)
+    # access data from remote source
+    region = cat.regionmask.MEOW.read()
+    print(region)
+
+    region.plot(add_label=False)
+
+    @savefig plotting_MEOW.png width=100%
+    plt.tight_layout()
```
Comment (mathause) [inline docs/intake_regionmask.rst:39]:
```suggestion
    @savefig plotting_MEOW.png width=100%
    region.plot(add_label=False)
```

File: docs/intake_regionmask.rst [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+You need to install intake_geopandas, which combines geopandas and intake, see
+https://intake.readthedocs.io/en/latest/.
+
+.. ipython:: python
+
+    import intake_geopandas
+    import intake
+    # open a pre-defined remote or local catalog yaml file
+    url = 'https://raw.githubusercontent.com/aaronspring/remote_climate_data/master/master.yaml'
```
Comment (mathause) [inline docs/intake_regionmask.rst:30]: Would it be difficult to have this catalog in regionmask itself? Then this docu would be independent of changes you do in your repo. You could add it to the docs folder...

File: docs/intake_regionmask.rst [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    cat = intake.open_catalog(url)
+    # access data from remote source
+    region = cat.regionmask.MEOW.read()
```
Comment (mathause) [inline docs/intake_regionmask.rst:33]:
```suggestion
    meow_regions = cat.regionmask.MEOW.read()
```
Could you give them a more descriptive name?

File: docs/intake_regionmask.rst [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    @savefig plotting_MEOW.png width=100%
+    plt.tight_layout()
+
+
+Build your own catalog
+======================
+
+Let's explore the Marine Ecoregions Of the World (MEOW) data set, which is a
+biogeographic classification of the world's coasts and shelves.
+To create a catalog we use the syntax described in
```
Comment (mathause) [inline docs/intake_regionmask.rst:47]: nit - could you switch the order of the sentences?

File: docs/whats_new.rst [hunk 8]
```diff
@@ -54,6 +54,8 @@ Docs
 
 - Unified the docstrings of all ``mask`` functions (:issue:`173`).
 - Mentioned how to calculate regional medians (:issue:`170`).
+- Mentioned how to open regions specified in a yaml file using intake and fsspec.
+ (:issue:`93`, :pull:`205` by `Aaron Spring <https://github.com/aaronspring>`_).
```
Comment (mathause) [inline docs/whats_new.rst:58]:
```suggestion
  (:issue:`93`, :pull:`205` by `Aaron Spring <https://github.com/aaronspring>`_).
```

File: docs/intake_regionmask.rst [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            abbrevs: _from_name
+            source: http://maps.tnc.org
+            numbers: ECO_CODE_X
+            name: MEOW
+    """)
+
+.. ipython:: python
+
+    cat = intake.open_catalog('test_cat.yml')
+    region = cat.MEOW.read()
```
Comment (mathause) [inline docs/intake_regionmask.rst:79]: Also here I would prefer another name.

**Review by mathause — COMMENTED** [review]
Some more nits. I think after this it's ready to go! The test failure is unrelated.

File: docs/intake_regionmask.rst (hunk 5 above)
Comment (aaronspring) [reply docs/intake_regionmask.rst:30]: sure. I could also just create a small catalog in the regionmask, but I want to still access it via http to show that remote catalogs are possible

Comment (aaronspring) [reply docs/intake_regionmask.rst:30]: created `data/regions_remote_catalog.yaml`

File: docs/intake_regionmask.rst (hunk 9 above)
Comment (aaronspring) [reply docs/intake_regionmask.rst:77]: `regions_my_local_catalog.yml` ?

**Comment by aaronspring** [comment]
It runs on my branch:
```
url = 'https://raw.githubusercontent.com/aaronspring/regionmask/AS_intake_docu/data/regions_remote_catalog.yaml'
cat = intake.open_catalog(url)
# access data from remote source
meow_regions = cat.MEOW.read()
```
but not yet on your master. but would after merge.

File: docs/whats_new.rst [hunk 10]
```diff
@@ -54,6 +54,8 @@ Docs
 
 - Unified the docstrings of all ``mask`` functions (:issue:`173`).
 - Mentioned how to calculate regional medians (:issue:`170`).
+- Mentioned how to open regions specified in a yaml file using intake and fsspec.
+  (:issue:`93`, :pull:`205` by `Aaron Spring <https://github.com/aaronspring>`_).
```
Comment (mathause) [inline docs/whats_new.rst:58]:
```suggestion
- Mentioned how to open regions specified in a yaml file using intake and fsspec
  (:issue:`93`, :pull:`205`). By `Aaron Spring <https://github.com/aaronspring>`_.
```

File: docs/intake_regionmask.rst [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+.. ipython:: python
+
+    cat = intake.open_catalog('regions_my_local_catalog.yml')
+    region = cat.MEOW.read()
+    print(region)
+
+
+Because ``simplecache::`` was added to the urlpath and ``use_fsspec=True``, the zip file was
+downloaded to the folder specified in cache_storage. The file access is now locally.
```
Comment (mathause) [inline docs/intake_regionmask.rst:84]:
```suggestion
downloaded to the folder specified in cache_storage. The file access is now local.
```

File: data/regions_remote_catalog.yaml [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    driver: intake_geopandas.regionmask.RegionmaskSource
+    args:
+      urlpath: simplecache::https://github.com/mathause/regionmask/raw/master/data/IPCC-WGI-reference-regions-v4.zip
+      use_fsspec: true
+      storage_options:
+        simplecache:
+          same_names: true
+      regionmask_kwargs:
+        names: Name
+        abbrevs: Acronym
```
Comment (mathause) [inline data/regions_remote_catalog.yaml:37]:
```suggestion
```

File: docs/intake_regionmask.rst [hunk 13]
```diff
@@ -0,0 +1,89 @@
+############################
+Using regionmask with intake
+############################
+
+Regions from geopandas shapefiles can be also pre-defined in a yaml file, which can be
```
Comment (mathause) [inline docs/intake_regionmask.rst:5]:
```suggestion
Regions from geopandas shapefiles can be pre-defined in a yaml file, which can be
```

File: docs/intake_regionmask.rst [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+            source: http://maps.tnc.org
+            numbers: ECO_CODE_X
+            name: MEOW
+    """)
+
+.. ipython:: python
+
+    cat = intake.open_catalog('regions_my_local_catalog.yml')
+    region = cat.MEOW.read()
+    print(region)
```
Comment (mathause) [inline docs/intake_regionmask.rst:80]:
```suggestion
    meow_regions = cat.MEOW.read()
    print(meow_regions)
```

**Comment by mathause** [comment]
:tada: thanks @aaronspring - it worked https://regionmask.readthedocs.io/en/latest/intake_regionmask.html

**Pull request merged**